s3datarepository/users.go

```diff
@@ -0,0 +1,447 @@
+package s3datarepository
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/YaleSpinup/ds-api/apierror"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+func (s *S3Repository) ListUsers(ctx context.Context, id string) (map[string]interface{}, error) {
+	if id == "" {
+		return nil, apierror.New(apierror.ErrBadRequest, "invalid input", errors.New("empty id"))
+	}
+
+	name := id
+	if s.NamePrefix != "" {
+		name = s.NamePrefix + "-" + name
+	}
+
+	log.Infof("listing users of the s3datarepository %s", name)
+
+	groupName := fmt.Sprintf("%s-DsTmpGrp", name)
+	usersOutput, err := s.listGroupsUsers(ctx, groupName)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Debugf("got iam users response %+v", usersOutput)
+
+	output := make(map[string]interface{}, len(usersOutput))
+	for _, u := range usersOutput {
+		keyOut, err := s.IAM.ListAccessKeysWithContext(ctx, &iam.ListAccessKeysInput{UserName: u.UserName})
+		if err != nil {
+			return nil, ErrCode("failed to getting access keys for user "+aws.StringValue(u.UserName), err)
+		}
+
+		keys := make(map[string]string, len(keyOut.AccessKeyMetadata))
+		for _, k := range keyOut.AccessKeyMetadata {
+			keys[aws.StringValue(k.AccessKeyId)] = aws.StringValue(k.Status)
+		}
+
+		output[*u.UserName] = struct {
+			Keys map[string]string `json:"keys"`
+		}{keys}
+	}
+
+	log.Debugf("returning map of users for dataset %s: %+v", id, output)
+
+	return output, nil
+}
+
+// listGroupsUsers lists the users that belong to a group
+func (s *S3Repository) listGroupsUsers(ctx context.Context, groupName string) ([]*iam.User, error) {
+	users := []*iam.User{}
+	if groupName == "" {
+		return users, apierror.New(apierror.ErrBadRequest, "invalid input", errors.New("empty group name"))
+	}
+
+	log.Infof("listing iam users for group %s", groupName)
+
+	input := &iam.GetGroupInput{GroupName: aws.String(groupName)}
+
+	truncated := true
+	for truncated {
+		output, err := s.IAM.GetGroupWithContext(ctx, input)
+		if err != nil {
+			return users, ErrCode("failed to getting users for group "+groupName, err)
+		}
+
+		truncated = aws.BoolValue(output.IsTruncated)
+		users = append(users, output.Users...)
+		input.Marker = output.Marker
+	}
+
+	return users, nil
+}
+
+func (s *S3Repository) CreateUser(ctx context.Context, id string) (interface{}, error) {
+	if id == "" {
+		return nil, apierror.New(apierror.ErrBadRequest, "invalid input", errors.New("empty id"))
+	}
+
+	name := id
+	if s.NamePrefix != "" {
+		name = s.NamePrefix + "-" + name
+	}
+
+	log.Infof("creating user of the s3datarepository %s", name)
```

Comment:
Oh, here's the `Info` 😄 So maybe it's fine to leave Debug in the handler, although should probably be the other way around 🤷

Reply:
Yeah i think we should standardize these maybe in our style guide.  I'm not 100% sure, but I actually think the Info should be in the called functions since then when you orchestrate a bunch of things you don't need a million Info log statements in the handler.

Reply:
Yeah, makes sense to keep the handler simple